### PR TITLE
Support Kerberos / GSSAPI in LibPQ on Linux

### PR DESCRIPTION
**Pull request by tp2750**
This adds support for Kerberos / gssapi authentication to LibPQ.jl on Linux.
See https://github.com/invenia/LibPQ.jl/issues/239)

Even though Kerberos_krb5 also builds for FreeBSD, the building of LibPQ failed for FreeBSD with --with-gssapi.

### CODE REVIEW

**Comment by giordano** [comment]
> Even though Kerberos_krb5 also builds for FreeBSD, the building of LibPQ failed for FreeBSD with --with-gssapi.

For reference, the error is
```
configure:12354: cc -o conftest -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Werror=unguarded-availability-new -Wendif-labels -Wmissing-format-attribute -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -Wno-unused-command-line-argument -O2   -I/workspace/destdir/include   -L/workspace/destdir/lib conftest.c -lgssapi -lkrb5 -lcrypto  -lpthread -lexecinfo -lm  >&5
/workspace/destdir/lib/libkrb5.so: undefined reference to `add_error_table'
/workspace/destdir/lib/libkrb5.so: undefined reference to `remove_error_table'
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
configure:12354: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "PostgreSQL"
| #define PACKAGE_TARNAME "postgresql"
| #define PACKAGE_VERSION "14.1"
| #define PACKAGE_STRING "PostgreSQL 14.1"
| #define PACKAGE_BUGREPORT "pgsql-bugs@lists.postgresql.org"
| #define PACKAGE_URL "https://www.postgresql.org/"
| #define CONFIGURE_ARGS " '--prefix=/workspace/destdir' '--build=x86_64-linux-musl' '--host=x86_64-unknown-freebsd12.2' '--with-includes=/workspace/destdir/include' '--with-libraries=/workspace/destdir/lib' '--without-readline' '--without-zlib' '--with-openssl' '--with-gssapi' 'build_alias=x86_64-linux-musl' 'host_alias=x86_64-unknown-freebsd12.2' 'CC=cc' 'CXX=c++' 'PKG_CONFIG_PATH=/workspace/destdir/lib/pkgconfig:/workspace/destdir/lib64/pkgconfig:/workspace/destdir/share/pkgconfig'"
| #define PG_MAJORVERSION "14"
| #define PG_MAJORVERSION_NUM 14
| #define PG_MINORVERSION_NUM 1
| #define PG_VERSION "14.1"
| #define DEF_PGPORT 5432
| #define DEF_PGPORT_STR "5432"
| #define BLCKSZ 8192
| #define RELSEG_SIZE 131072
| #define XLOG_BLCKSZ 8192
| #define ENABLE_THREAD_SAFETY 1
| #define ENABLE_GSS 1
| #define PG_KRB_SRVNAM "postgres"
| #define STDC_HEADERS 1
| #define HAVE_SYS_TYPES_H 1
| #define HAVE_SYS_STAT_H 1
| #define HAVE_STDLIB_H 1
| #define HAVE_STRING_H 1
| #define HAVE_MEMORY_H 1
| #define HAVE_STRINGS_H 1
| #define HAVE_INTTYPES_H 1
| #define HAVE_STDINT_H 1
| #define HAVE_UNISTD_H 1
| #define HAVE_PTHREAD_PRIO_INHERIT 1
| #define HAVE_PTHREAD 1
| #define HAVE_STRERROR_R 1
| #define HAVE_GETPWUID_R 1
| #define HAVE_GETHOSTBYNAME_R 1
| #define STRERROR_R_INT 1
| #define HAVE_LIBM 1
| #define HAVE_SPINLOCKS 1
| #define HAVE_ATOMICS 1
| /* end confdefs.h.  */
| 
| /* Override any GCC internal prototype to avoid an error.
|    Use char because int might match the return type of a GCC
|    builtin and then its argument prototype would still apply.  */
| #ifdef __cplusplus
| extern "C"
| #endif
| char gss_init_sec_context ();
| int
| main ()
| {
| return gss_init_sec_context ();
|   ;
|   return 0;
| }
configure:12371: result: no
configure:12378: error: could not find function 'gss_init_sec_context' required for GSSAPI
```
But I don't understand it:
```console
sandbox:${WORKSPACE}/srcdir/postgresql-14.1 # readelf -d ${libdir}/libkrb5.so | head -n8

Dynamic section at offset 0x341000 contains 21 entries:
  Tag        Type                         Name/Value
 0x000000000000000e (SONAME)             Library soname: [libkrb5.so.3]
 0x0000000000000001 (NEEDED)             Shared library: [libk5crypto.so]
 0x0000000000000001 (NEEDED)             Shared library: [libcom_err.so]
 0x0000000000000001 (NEEDED)             Shared library: [libkrb5support.so]
 0x000000000000001d (RUNPATH)            Library runpath: [$ORIGIN]
sandbox:${WORKSPACE}/srcdir/postgresql-14.1 # nm ${libdir}/libkrb5.so | grep add_error_table
                 U add_error_table
sandbox:${WORKSPACE}/srcdir/postgresql-14.1 # nm ${libdir}/libcom_err.so | grep add_error_table
0000000000001130 T add_error_table
```
The symbol is undefined in `libkrb5.so`, but it's provided (and exported) by `libcom_err.so`, which is required by `libkrb5.so` and it's in the same directory (and `$ORIGIN` is in the runpath).

File: L/LibPQ/build_tarballs.jl [hunk 1]
```diff
@@ -34,8 +34,13 @@ if [[ "${target}" == i686-linux-musl ]]; then
     # other tests.
     sed -i 's/cross_compiling=no/cross_compiling=yes/' configure
 fi
-./configure --prefix=${prefix} --build=${MACHTYPE} --host=${target} --with-includes=$prefix/include --with-libraries=$prefix/lib --without-readline --without-zlib --with-openssl
+if [[ "${target}" == *"linux"* ]] ; then
+    ./configure --prefix=${prefix} --build=${MACHTYPE} --host=${target} --with-includes=$prefix/include --with-libraries=$prefix/lib --without-readline --without-zlib --with-openssl --with-gssapi
+ else
+    ./configure --prefix=${prefix} --build=${MACHTYPE} --host=${target} --with-includes=$prefix/include --with-libraries=$prefix/lib --without-readline --without-zlib --with-openssl
+ fi
```
Comment (giordano) [inline L/LibPQ/build_tarballs.jl:41]: I don't like long duplicate lists of flags.
```suggestion
FLAGS=()
if [[ "${target}" == *-linux-* ]]; then
    FLAGS+=(--with-gssapi)
fi
./configure --prefix=${prefix} \
    --build=${MACHTYPE} \
    --host=${target} \
    --with-includes=${includedir} \
    --with-libraries=${libdir} \
    --without-readline \
    --without-zlib \
    --with-openssl \
    "${FLAGS[@]}"
```

Comment (giordano) [reply L/LibPQ/build_tarballs.jl]: I pushed these changes, together with the fix for FreeBSD

Comment (tp2750) [reply L/LibPQ/build_tarballs.jl]: Thanks a lot.

Great you found out what was the problem for FreeBSD.

**Pull request merged**